ovn-tester/cms/openstack/openstack.py

```diff
@@ -0,0 +1,561 @@
+import logging
+import uuid
+from dataclasses import dataclass
+from typing import List, Optional, Dict
+
+import netaddr
+
+from randmac import RandMac
+
+from ovn_context import Context
+from ovn_utils import (
+    DualStackSubnet,
+    LSwitch,
+    LSPort,
+    PortGroup,
+    LRouter,
+    LRPort,
+    DualStackIP,
+)
+from ovn_workload import ChassisNode, Cluster
+
+log = logging.getLogger(__name__)
+
+OVN_HEATER_CMS_PLUGIN = 'Openstack'
+
+
+@dataclass
+class NeutronNetwork:
+    """Group of OVN objects that logically belong to same Neutron network."""
+
+    network: LSwitch
+    ports: Dict[str, LSPort]
+    security_group: Optional[PortGroup] = None
+
+    def __post_init__(self):
+        self._ip_pool = self.network.cidr.iter_hosts()
+        self.gateway = self.next_host_ip()
+
+    def next_host_ip(self) -> netaddr.IPAddress:
+        """Return next available host IP in this network's range."""
+        return next(self._ip_pool)
+
+
+class Project:
+    """Represent network components of an OpenStack Project aka. Tenant."""
+
+    def __init__(
+        self,
+        networks: List[NeutronNetwork] = None,
+        router: Optional[LRouter] = None,
+    ):
+        self.networks: List[NeutronNetwork] = (
+            [] if networks is None else networks
+        )
+        self.router = router
+        self._id = str(uuid.uuid4())
+
+    @property
+    def uuid(self) -> str:
+        """Return arbitrary UUID assigned to this Openstack Project."""
+        return self._id
+
+
+class OpenStackCloud(Cluster):
+    """Representation of Openstack cloud/deployment."""
+
+    def __init__(self, central_node, worker_nodes, cluster_cfg, brex_cfg):
+        super().__init__(central_node, worker_nodes, cluster_cfg, brex_cfg)
+        self.router = None
+        self.external_port: Optional[LRPort] = None
+        self._projects: List[Project] = []
+
+        self._int_net_base = DualStackSubnet(cluster_cfg.node_net)
+        self._int_net_offset = 0
+
+        self._ext_net_pool_index = 0
+
+    @property
+    def projects(self) -> List[Project]:
+        """Return list of Projects that exist in Openstack deployment."""
+        return self._projects
+
+    def next_external_ip(self) -> DualStackIP:
+        """Return next available IP address from configured 'external_net'."""
+        self._ext_net_pool_index += 1
+        return self.cluster_cfg.external_net.forward(self._ext_net_pool_index)
+
+    def next_int_net(self):
+        """Return next subnet for project/tenant internal network.
+
+        This method takes subnet configured as 'node_net' in cluster config as
```

Comment:
Note that I think many OpenStack clouds default to re-using the same CIDR for the internal network of every project.

Using different ones makes it simpler to have overview in a test though and allows for testing both NAT'ed and routed use cases.

I guess we need to consider whether having many networks using the same CIDR impacts scale and whether we need to test that somewhere down the line.

Reply:
Agreed, I think your approach is the best to start with as it is easier to work with and covers multiple use cases. If we come to the conclusion that creating multiple networks with the same CIDR is an important use case we can change it later.